draft-irtf-cfrg-hpke.md

```diff
@@ -794,35 +786,80 @@ def OpenAuthPSK(enc, skR, info, aad, ct, psk, pskID, pkS):
 
 ## Key Encapsulation Mechanisms (KEMs) {#kem-ids}
 
-| Value  | KEM                            | Nzz  | Nenc | Npk | Reference                    |
-|:-------|:-------------------------------|:-----|:-----|:----|:-----------------------------|
-| 0x0000 | (reserved)                     | N/A  | N/A  | N/A | N/A                          |
-| 0x0010 | DHKEM(P-256, HKDF-SHA256)      | 32   | 65   | 65  | {{NISTCurves}}, {{?RFC5869}} |
-| 0x0011 | DHKEM(P-384, HKDF-SHA384)      | 48   | 97   | 97  | {{NISTCurves}}, {{?RFC5869}} |
-| 0x0012 | DHKEM(P-521, HKDF-SHA512)      | 64   | 133  | 133 | {{NISTCurves}}, {{?RFC5869}} |
-| 0x0020 | DHKEM(X25519, HKDF-SHA256) | 32   | 32   | 32  | {{?RFC7748}}, {{?RFC5869}}   |
-| 0x0021 | DHKEM(X448, HKDF-SHA512)   | 64   | 56   | 56  | {{?RFC7748}}, {{?RFC5869}}   |
+| Value  | KEM                            | Nzz  | Nenc | Npk | Nsk | Reference                    |
+|:-------|:-------------------------------|:-----|:-----|:----|:----|:-----------------------------|
+| 0x0000 | (reserved)                     | N/A  | N/A  | N/A | N/A | N/A                          |
+| 0x0010 | DHKEM(P-256, HKDF-SHA256)      | 32   | 65   | 65  | 32  | {{NISTCurves}}, {{?RFC5869}} |
+| 0x0011 | DHKEM(P-384, HKDF-SHA384)      | 48   | 97   | 97  | 48  | {{NISTCurves}}, {{?RFC5869}} |
+| 0x0012 | DHKEM(P-521, HKDF-SHA512)      | 64   | 133  | 133 | 66  | {{NISTCurves}}, {{?RFC5869}} |
+| 0x0020 | DHKEM(X25519, HKDF-SHA256) | 32   | 32   | 32  | 32  | {{?RFC7748}}, {{?RFC5869}}   |
+| 0x0021 | DHKEM(X448, HKDF-SHA512)   | 64   | 56   | 56  | 56  | {{?RFC7748}}, {{?RFC5869}}   |
+
+### Marshal/Unmarshal
+
+For the NIST curves P-256, P-384 and P-521, the Marshal function of the
+KEM performs the uncompressed Elliptic-Curve-Point-to-Octet-String
+conversion according to {{SECG}}. The Unmarshal function performs the
+uncompressed Octet-String-to-Elliptic-Curve-Point conversion.
+
+For the CFRG curves Curve25519 and Curve448, the Marshal and Unmarshal functions
+are the identity function, since these curves already use fixed-length byte
+strings for public keys.
+
+Some unmarshalled public keys MUST be validated before they can be used. See
+{{validation}} for specifics.
+
+### DeriveKeyPair
+
+For the NIST curves P-256, P-384 and P-521, the DeriveKeyPair function of the
+KEM performs rejection sampling over field elements:
 
-### Marshal
+~~~
+def DeriveKeyPair(ikm):
+  prk = LabeledExtract(zero(0), "nistp_keypair", ikm)
+  sk = 0
+  counter = 1
+  while sk < 1 or order <= sk:
+    label = concat("candidate ", encode_big_endian(counter, 1))
+    bytes = Expand(prk, label, Nsk)
+    bytes[0] = bytes[0] & bitmask
+    sk = decode_big_endian(bytes)
+    counter = counter + 1
+  return (sk, pk(sk))
+~~~
+
+where `order` is the order of the curve being used (this can be found in
+section D.1.2 of {{NISTCurves}}), and `bitmask` is defined to be 0xFF for P-256
+and P-384, and 0x01 for P-521.
```

Comment:
Where does this value come from?

Reply:
This is because the scalar size for P-521 is 521 bits, or 65 bytes plus 1 bit. In order to get a scalar in this range, we have to generate 66 bytes and clear the top 7 bits. Otherwise the rejection sampling will fail each iteration with probability 254/255.